<h1 style="text-align: center;">Methods of Inspection</h1>

## Line profiler

line_profiler is a Python module for analyzing code execution time, especially for each line of code execution time in detail. It can help us locate performance bottlenecks in our code and find areas that can be optimized. By using line_profiler, we can better understand the performance characteristics of the code and target optimizations to improve the efficiency of program execution.

### Install

Run the command in the terminal:`pip install line-profiler`

### Example

Let's take a simple example: calculating the sum of squares of a large array.

In [48]:
large_arr = np.random.randint(1, 100, size=100000)

def sum_of_squares(arr):
    result = 0
    for num in arr:
        result += num ** 2
    return result

import line_profiler
profiler = line_profiler.LineProfiler()
profiler.add_function(sum_of_squares)
profiler.run("sum_of_squares(large_arr)")
profiler.print_stats()

Timer unit: 1e-07 s

Total time: 0.105808 s
File: C:\Users\DELL\AppData\Local\Temp\ipykernel_15492\2110549466.py
Function: sum_of_squares at line 3

Line #      Hits         Time  Per Hit   % Time  Line Contents
     3                                           def sum_of_squares(arr):
     4         1          4.0      4.0      0.0      result = 0
     5    100000     406596.0      4.1     38.4      for num in arr:
     6    100000     651475.0      6.5     61.6          result += num ** 2
     7         1          3.0      3.0      0.0      return result



`Line #`: The line number in the code.

`Hits`: The number of times the code was executed.

`Time`: Total time. Indicates the total time (in seconds) accumulated in all executions of the code.

`Per Hit`: Average Time. Represents the average time (Time/Hits) of each execution of the code.

`% Time`: Percentage time. Indicates the percentage of time spent running the code out of the total program runtime.


In the output, we can see that for the `sum_of_squares` function, each line represents the corresponding line of code in the function (from 3 to 7 lines) and lists the number of times each line of code was executed, the total elapsed time, the average elapsed time, and the percentage time.

For example, in line 5 of the code, the line `for num in arr`: was executed 100,000 times at runtime, taking a total of 406,596 microseconds (i.e., 0.406596 seconds), with an average of 4.1 microseconds (i.e., 0.0000041 seconds), and occupying 38.4% of the total runtime.

In this case, the `result += num ** 2` in the for loop takes up the vast majority of the time. This suggests that when working with large-scale data, we may be able to get better performance by using NumPy's vectorization operation instead of the for loop: `result = np.sum(arr ** 2)`.

Therefore, using `line_profiler` can help us locate performance bottlenecks and identify potential optimization points to improve code execution efficiency.

## View check

当创建一个数组的视图（view）时，它与原始数组共享相同的数据缓冲区。通常情况下，使用view更快，因为它避免了数据的复制和额外的内存分配。因此我们有检查view的必要。

### Example

我们首先创建一个数组C。

In [49]:
import numpy as np
C = np.arange(12)
C = np.arange(12).reshape(4,3)
C

array([[ 0,  1,  2],
       [ 3,  4,  5],
       [ 6,  7,  8],
       [ 9, 10, 11]])

使x等于C数组的第一行：

In [50]:
x = C[0,:]
x

array([0, 1, 2])

当我们改变数组的第一个元素时，x也会相应改变，因为x是view。

In [51]:
C[0,0] = 100
C

array([[100,   1,   2],
       [  3,   4,   5],
       [  6,   7,   8],
       [  9,  10,  11]])

In [52]:
x

array([100,   1,   2])

比如下面这个例子，我们首先创建一个NumPy数组。

In [93]:
import numpy as np

data_list = [x**2 for x in range(1000**2)]
data_np = np.array(data_list)

接着计算平方根。这里我们计算一下运算时间。

In [94]:
start_time = time.time()
result = np.sqrt(data_np)
end_time = time.time()
print(f"execution time: {end_time - start_time} seconds")

NumPy execution time: 0.007866621017456055 seconds


我们可以用`.base`来检查一下，data_np是不是view。

In [95]:
print(data_np.base)

None


结果显示它并不是，因此这里我们可以为它创建一个视图，并重新计算平方和。

In [99]:
start_time = time.time()
data_np_view = data_np.view()
result = np.sqrt(data_np_view)
end_time = time.time()
print(f"execution time: {end_time - start_time} seconds")

execution time: 0.00401759147644043 seconds
Timer unit: 1e-07 s



In [97]:
print(data_np_view.base)

[           0            1            4 ... 999994000009 999996000004
 999998000001]


In [98]:
np.shares_memory(data_np_view,  data_np)

True